### PR TITLE
feat(flagd-proxy): introduce zero-downtime

### LINKED ISSUE #751: [FEATURE] flagd-proxy: support zero downtime upgrades
## Goal

Flagd-proxy shall be upgraded without any downtime and run with multiple replicas.

### Details

Flagd-proxy should be able to handle upgrades without any downtime.
For this, we should make sure that it handles the [pre-stop signals of K8s](https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/).
When the SIGTERM and SIGINT signals are sent to the container, Flagd-proxy should close any open connection with all FlagD-sidecar.
FlagD-Sidecar should try to reconnect again and should hit a ready replica of Flagd-proxy
Furthermore, Flagd-proxy should run as PID 1 in the container, otherwise it won't receive the signals emitted by K8s.

### DoD
Flagd-proxy runs as PID1
Flagd-proxy handles termination gracefully
FlagD-Sidecar tries to reconnect if a Flagd-proxy is upgraded
Readiness probe is "disabled" as soon as the termination signal is received
Tests to cover this behavior are provided

### Resources
K8s prestop hook: https://kubernetes.io/docs/concepts/containers/container-lifecycle-hooks/
PID1: https://about.gitlab.com/blog/2022/05/17/how-we-removed-all-502-errors-by-caring-about-pid-1-in-kubernetes/

### PR DESCRIPTION
**Pull request by odubajDT**
### Changes

 - implements graceful shutdown of the server
 - create grpc client that fails, when it's not able to re-connect to the server -> flagd-proxy server is unavailable
 - create manifests for deploying flagd-proxy and pod, which runs the client
 - create bash script that runs zero-downtime test
 - create README to describe how to build and run zero-downtime test
 
 ### Test description
First the environment (flagd-proxy and local flag config) is deployed. Afterwards a pod with the grpc client is deployed on the k8s cluster, where the client is configured to fail (and with that the whole pod) when it's not able to connect to the flagd-proxy server (for example after update). If the client is still running at the end and it has 0 restarts (every re-connection was ok) -> no connection problem to the flagd-proxy server -> test passes.

During this time (when the client/pod is running), there are multiple upgrades (6) of flagd-proxy server. This together tests that the flagd-proxy server is always available for the client to re-connect.

### Follow-up
This test can be fully automated and could be part of the zero-downtime pipeline. Adaptation of the pipeline will be part of a separate PR if the team decides to do it.

### Related Issues

Fixes #751

### CODE REVIEW

File: test/zero-downtime-flagd-proxy/README.md [hunk 1]
```diff
@@ -0,0 +1,28 @@
+# FlagDT Zero downtime test
```
Comment (beeme1mr) [inline test/zero-downtime-flagd-proxy/README.md:1]:
```suggestion
# FlagD Proxy Zero downtime test
```

File: test/zero-downtime-flagd-proxy/main.go [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+func main() {
+	waitSecondsBetweenRequests := getWaitSecondsBetweenRequests()
+	flagDTURL := getURL()
+
+	// Create a channel to receive a signal when the gRPC connection fails
+	errChan := make(chan bool)
+
+	// Use a goroutine to run your program logic
+	go func() {
+		if err := handleRequests(waitSecondsBetweenRequests, flagDTURL); err != nil {
```
Comment (beeme1mr) [inline test/zero-downtime-flagd-proxy/main.go:21]:
```suggestion
		if err := handleRequests(waitSecondsBetweenRequests, flagdURL); err != nil {
```

File: test/zero-downtime-flagd-proxy/README.md [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+## How to run
+
+Clone this repository and run the following command:
+
+```shell
+FLAGD_PROXY_IMG=your-flagdt-image FLAGD_PROXY_IMG_ZD=your-flagdt-second-image ZD_CLIENT_IMG=your-zd-client-image make run-zd-test
```
Comment (beeme1mr) [inline test/zero-downtime-flagd-proxy/README.md:8]:
```suggestion
FLAGD_PROXY_IMG=your-flagd-image FLAGD_PROXY_IMG_ZD=your-flagd-second-image ZD_CLIENT_IMG=your-zd-client-image make run-zd-test
```

File: test/zero-downtime-flagd-proxy/README.md [hunk 4]
```diff
@@ -0,0 +1,28 @@
+# FlagD Proxy Zero downtime test
+
+## How to run
+
+Clone this repository and run the following command:
+
+```shell
+FLAGD_PROXY_IMG=your-flagd-image FLAGD_PROXY_IMG_ZD=your-flagd-second-image ZD_CLIENT_IMG=your-zd-client-image make run-zd-test
```
Comment (toddbaert) [inline test/zero-downtime-flagd-proxy/README.md:8]:
```suggestion
FLAGD_PROXY_IMG=your-flagd-proxy-image FLAGD_PROXY_IMG_ZD=your-flagd-proxy-second-image ZD_CLIENT_IMG=your-zd-client-image make run-zd-test
```

File: test/zero-downtime-flagd-proxy/README.md [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+```
+
+This will create a flagd-proxy and a job in `flagd-zd-test` namespace,
+where the test will be run.
+
+Please be aware, you need to build your custom image for the zd-client
+and two images for flagD first.
```
Comment (toddbaert) [inline test/zero-downtime-flagd-proxy/README.md:15]: I'm a bit confused here. I know how to build my flagd proxy images, but do I build the zd-client image? I don't see a Dockerfile for it.

I'm guessing there's also other prerequisites here? Like having a cluster available, and the images above need to be available to it (ie: a local kind cluster won't have locally built containers available to it).

File: test/zero-downtime-flagd-proxy/README.md [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+
+To build your images using [ko](https://github.com/ko-build/ko),
+you need to login to your repository, where the images will be pushed:
```
[the remark this answers is not in the document]
Comment (odubajDT) [reply test/zero-downtime-flagd-proxy/README.md:18]: We can surely create a Dockerfile, it will be very simple, but since it's not a "production software image" I choose `ko` as it does not need  any other additional file to build and it's quite straight-forward to use. I might think that in the future we will adapt also the rest of the images, as it has some advantages to docker. You can find them [here](https://skaffold.dev/docs/builders/builder-types/ko/). 

But yes, for now we can still stick to docker and I can provide a Dockerfile :)

File: core/pkg/service/sync/server.go [hunk 7]
```diff
@@ -18,11 +18,13 @@ import (
...
-	Logger        *logger.Logger
-	handler       *handler
-	config        iservice.Configuration
+	server           *http.Server
+	metricsServer    *http.Server
+	Logger           *logger.Logger
+	handler          *handler
+	config           iservice.Configuration
+	grpcServer       *grpc.Server
+	readinessEnabled bool
```
Comment (thisthat) [inline core/pkg/service/sync/server.go:27]: I suggest changing the name since `readinessEnabled` convey a different purpose than what is used for in the code 
```suggestion
	serverReady bool
```

File: flagd-proxy/build.Dockerfile [hunk 8]
```diff
@@ -32,13 +32,11 @@ RUN --mount=type=cache,target=/go/pkg/mod/ \
     --mount=type=bind,source=./core,target=./core \
     --mount=type=bind,source=./flagd,target=./flagd \
     --mount=type=bind,source=./flagd-proxy,target=./flagd-proxy \
-    CGO_ENABLED=0 GOOS=${TARGETOS} GOARCH=${TARGETARCH} go build -a -ldflags "-X main.version=${VERSION} -X main.commit=${COMMIT} -X main.date=${DATE}" -o /bin/flagd-build flagd-proxy/main.go
+    CGO_ENABLED=0 GOOS=${TARGETOS} GOARCH=${TARGETARCH} go build -a -ldflags "-X main.version=${VERSION} -X main.commit=${COMMIT} -X main.date=${DATE}" -o /bin/flagd-proxy flagd-proxy/main.go
 
-# # Use distroless as minimal base image to package the manager binary
-# # Refer to https://github.com/GoogleContainerTools/distroless for more details
-FROM gcr.io/distroless/static:nonroot
+FROM alpine:3.18
```
Comment (thisthat) [inline flagd-proxy/build.Dockerfile:37]: I would revert the change and use the distroless image

Comment (odubajDT) [reply flagd-proxy/build.Dockerfile:37]: I was unable to make it work with the distroless image, not sure why... I can examine it further.

File: core/pkg/service/sync/server.go (hunk 7 above)
Comment (odubajDT) [reply core/pkg/service/sync/server.go:27]: I do not find `serverReady` as a good name here.. I would like to stick to the readiness enabled

Comment (thisthat) [reply core/pkg/service/sync/server.go:27]: you're right, this is only for the metric part. Maybe `metricServerReady` is the most descriptive name since also not all traffic might be allowed later on

File: test/zero-downtime-flagd-proxy/main.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+				fmt.Println(err.Error())
+			}
+		}
+	}()
+
+	return doRequests(grpcClient, waitSecondsBetweenRequests)
+}
+
+func getWaitSecondsBetweenRequests() int {
+	return getEnvVarOrDefault("WAIT_TIME_BETWEEN_REQUESTS_MS", 1)
```
Comment (bacherfl) [inline test/zero-downtime-flagd-proxy/main.go:50]: here we have a slight mismatch, as the name of the env var suggests that the wait time is provided in milliseconds, however that value is treaded as seconds in the code

File: test/zero-downtime-flagd-proxy/zd_test.sh [hunk 10]
```diff
@@ -0,0 +1,56 @@
...
+kubectl create ns $ZD_TEST_NAMESPACE_FLAGD_PROXY
+envsubst < test/zero-downtime-flagd-proxy/manifests/proxy/deployment.yaml | kubectl apply -f - -n $ZD_TEST_NAMESPACE_FLAGD_PROXY
+kubectl apply -f test/zero-downtime-flagd-proxy/manifests/proxy/service.yaml -n $ZD_TEST_NAMESPACE_FLAGD_PROXY
+kubectl apply -f test/zero-downtime-flagd-proxy/manifests/proxy/flag-config.yaml -n $ZD_TEST_NAMESPACE_FLAGD_PROXY
+kubectl wait --for=condition=available deployment/flagd-proxy -n $ZD_TEST_NAMESPACE_FLAGD_PROXY --timeout=60s
+envsubst < test/zero-downtime-flagd-proxy/manifests/pod.yaml | kubectl apply -f - -n $ZD_TEST_NAMESPACE_FLAGD_PROXY
+kubectl wait --for=condition=ready pod/zd-test -n $ZD_TEST_NAMESPACE_FLAGD_PROXY --timeout=30s
+
+# Wait until connections from client to flagd-proxy are established
+sleep 20
```
Comment (bacherfl) [inline test/zero-downtime-flagd-proxy/zd_test.sh:19]: is that sleep still required, as we wait for the `zd-test` pod to be ready in the previous command?

File: test/zero-downtime-flagd-proxy/zd_test.sh [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+    # 20s should be enough, as terminationGracePeriod is set to 10s
+    sleep 20
+done
+
+# Pod will fail only when it fails to re-connect (that means we do not have zero downtime)
+# If it is still running, the last re-connection was successfull.
+kubectl wait --for=condition=ready pod/zd-test -n $ZD_TEST_NAMESPACE_FLAGD_PROXY --timeout=30s
+
+# If re-connection was once not successful and another re-connection was, pod might be in a ready state again.
+# Therefore we need to check that the restrt count is equal to zero -> this means every re-connection was ok.
```
Comment (bacherfl) [inline test/zero-downtime-flagd-proxy/zd_test.sh:47]:
```suggestion
# Therefore we need to check that the restart count is equal to zero -> this means every re-connection was ok.
```

File: test/zero-downtime-flagd-proxy/zd_test.sh [hunk 12]
```diff
@@ -0,0 +1,56 @@
...
+restart_count=$(kubectl get pods zd-test -o=jsonpath='{.status.containerStatuses[0].restartCount}' -n $ZD_TEST_NAMESPACE_FLAGD_PROXY)
+if [ "$restart_count" -ne 0 ]; then
+    echo "Restart count of the zd-test pod is not equal to zero."
```
Comment (bacherfl) [inline test/zero-downtime-flagd-proxy/zd_test.sh:50]: let's also print how many times it was restarted here

File: test/zero-downtime-flagd-proxy/zd_test.sh (hunk 10 above)
Comment (odubajDT) [reply test/zero-downtime-flagd-proxy/zd_test.sh:19]: It's safe to have it here, as the client still needs to connect to the flagd-proxy after it is ready. To be sure we do not update the Deployment of flagd-proxy before the client is connected, we should wait some time here

Comment (bacherfl) [reply test/zero-downtime-flagd-proxy/zd_test.sh:19]: Ok I was just thinking that this value might be overly cautious - maybe as a follow up, and when that test is executed regularly in the pipeline we will be able to reduce it

**Review by bacherfl — APPROVED** [review]

**Review by toddbaert — APPROVED** [review]

**Review by beeme1mr — APPROVED** [review]

**Review by thisthat — APPROVED** [review]
